Analytics docs example queries (#11657)
RasaHQ/rasa: docs/docs/monitoring/analytics/example-queries.mdx

---
id: example-queries
sidebar_label: Example queries
title: Example queries
description:
---

> 

Author: TODO: add MDX comment to explain how to get the graphs from https://rasa-technologies-gmbh.metabaseapp.com/dashboard/3-example-queries-for-the-docs

import useBaseUrl from "@docusaurus/useBaseUrl";

> Below you can find inspiration on how to analyze the conversations your assistant is having.

Reviewer: ```suggestion
This section aims to help you get started with ways on how to analyze the conversations your assistant is having.
```

> Example queries using SQL are provided, together with example visualisation in Metabase.

Reviewer: ```suggestion
Example queries using SQL are provided, together with an example visualisation in Metabase.
```

### Number of sessions per month

> It's common to look at high level usage metrics of your chatbot. A metric you could look at is number of

Reviewer: ```suggestion
A common high level usage metric of your assistant you could look at is the number of
```

sessions per month. Here is how it could look like:

```sql
SELECT
  date_trunc('month', "public"."rasa_session"."timestamp") AS "first_seen",
  count(*) AS "count"
FROM "public"."rasa_session"
GROUP BY 1
ORDER BY 1 ASC
```

<div align="center">
  <img
    alt="Example result for a query showcasing the number of users per day."
    src={useBaseUrl("/img/analytics/graph-number-sessions-month.png")}
    width="100%"
  />
</div>

### Number of sessions per channel

> If you're connecting your assistant to multiple channels, another metric that could be interesting to look at
> is the number of sessions per channel, let's say per week. You'd get the following query:

Reviewer: ```suggestion
If you're connecting your assistant to multiple channels, it could be useful to look at
the number of sessions per channel, let's say per week. The query you would need for this metric is:
```

```sql
SELECT
  "public"."rasa_user_message"."channel" AS "channel",
  "public"."rasa_user_message"."timestamp" AS "timestamp",
  count(distinct "public"."rasa_user_message"."session_id") AS "count"
FROM "public"."rasa_user_message"
GROUP BY 1, 2
ORDER BY 1 ASC, 2 ASC
```

<div align="center">
  <img
    alt="Example result for a query showcasing the number of users per day."
    src={useBaseUrl("/img/analytics/graph-number-sessions-channel.png")}
    width="100%"
  />
</div>

### Top N intents

> When improving your assistant, it's common to look at variety of intents your users express. The Top 5 intents metric

Reviewer 2: "the variety"

Reviewer 2: "The top 5 intent metrics" instead of "intents metric"

Author: d161380a4a5

> below will help you have a good perspective on that topic:

Reviewer: ```suggestion
To improve your assistant, you could looking into the variety of intents your users express. The query below selects the top 5 intents which could help you have a good perspective on that topic:
```

```sql
SELECT
  "public"."rasa_user_message"."intent" AS "intent",
  count(*) AS "count"
FROM "public"."rasa_user_message"
GROUP BY 1
ORDER BY 2 DESC, 1 ASC
LIMIT 5
```

<div align="center">
  <img
    alt="Example result for a query showcasing the number of users per day."
    src={useBaseUrl("/img/analytics/graph-top-5-intents.png")}
    width="100%"
  />
</div>

> You could even go further and look for the intent distribution over time:

Reviewer: ```suggestion
Moreover, you can look for the intent distribution over time:
```

```sql
SELECT
  "public"."rasa_user_message"."intent" AS "intent",
  date_trunc('month', "public"."rasa_user_message"."timestamp") AS "timestamp",
  count(*) AS "count" FROM "public"."rasa_user_message"
GROUP BY 1, 2
ORDER BY 1 ASC, 2 ASC
```

<div align="center">
  <img
    alt="Example result for a query showcasing the number of users per day."
    src={useBaseUrl("/img/analytics/graph-intent-distribution.png")}
    width="100%"
  />
</div>

### Escalation rate

> Going further in analysing what happens during a conversation, you might want to understand what's your escalation rate
> or human handoff rate. Let's say you have an intent named `handoff_to_support`. You'll get the escalation rate over time

Reviewer: A definition of this metric is needed for setting the context.

```suggestion
Escalation rate or human hand-off rate is a measure of how many conversations does the assistant pass to a human agent. This metric can help you gain a better understanding of what happens during a conversation. Let's say you have an intent named `handoff_to_support`. You'll get the escalation rate over time
```

with this sample query:

```sql
WITH "sessions" AS (
    SELECT
        "public"."rasa_user_message"."session_id" AS "session_id",
        date_trunc('month', "public"."rasa_user_message"."timestamp") AS "timestamp",
        (
          CASE "public"."rasa_user_message"."intent"
            WHEN 'handoff_to_support'
            THEN 1 ELSE 0
          END
        ) AS "has_handoff_to_support"
    FROM "public"."rasa_user_message"
),
"sessions_with_handoff" AS (
    SELECT
      "session_id",
      "timestamp",
      SUM(
        CASE
          WHEN "has_handoff_to_support" > 0
          THEN 1 ELSE 0
        END
      ) AS "has_handoff_to_support"
    FROM "sessions"
    GROUP BY 1, 2
)
SELECT
  "timestamp",
  SUM("has_handoff_to_support") / count(*) AS "escalation_rate"
FROM "sessions_with_handoff"
GROUP BY 1 ASC
ORDER BY 1 ASC
```

<div align="center">
  <img
    alt="Example result for a query showcasing the number of users per day."
    src={useBaseUrl("/img/analytics/graph-escalation-rate.png")}
    width="100%"
  />
</div>

### Abandonment rate

> Another interesting conversational metric to look at is the abandonment rate. You might have your own definition of what qualifies
> as abandonment of a conversation, but for simplicity here we'll define it as a session ending without a user message after
> a specific message was utterred by the bot, e.g. `utter_ask_name`.
> You could of course adapt the metric to detect sessions ending without a user message after a specific set of intents.

Reviewer: ```suggestion
Abandonment rate can be defined in many different custom ways, however here we'll define it as a session ending without a user message after a specific message was uttered by the bot, e.g. `utter_ask_name`.
You could adapt the metric to detect sessions ending without a user message after a specific set of intents.
```

> You could of course adapt the metric to detect sessions ending without a user message after a specific set of intents.

Reviewer 2: ```suggestion
You could also adapt the metric to detect sessions ending without a user message after a specific set of intents.
```

id: example-queries
sidebar_label: Example queries
title: Example queries
description:
---

import useBaseUrl from "@docusaurus/useBaseUrl";

This section aims to help you get started with ways on how to analyze the conversations your assistant is having.
Example queries using SQL are provided, together with an example visualisation in Metabase.

## Number of sessions per month

A common high level usage metric of your assistant you could look at is the number of
sessions per month. Here is how it would look:

```sql
SELECT
  date_trunc('month', "public"."rasa_session"."timestamp") AS "first_seen",
  count(*) AS "count"
FROM "public"."rasa_session"
GROUP BY 1
ORDER BY 1 ASC
```

<figure align="center">
  <img
    alt="Number of sessions per month visualised in Metabase."
    src={useBaseUrl("/img/analytics/graph-number-sessions-month.png")}
    width="100%"
  />
  <figcaption>Number of sessions per month visualised in Metabase.</figcaption>
</figure>

## Number of sessions per channel

If you're connecting your assistant to multiple channels, it could be useful to look at
the number of sessions per channel, let's say per week. The query you would need for this metric is:

```sql

> SELECT
>   "public"."rasa_user_message"."channel" AS "channel",
>   "public"."rasa_user_message"."timestamp" AS "timestamp",
>   count(distinct "public"."rasa_user_message"."session_id") AS "count"
> FROM "public"."rasa_user_message"
> GROUP BY 1, 2
> ORDER BY 1 ASC, 2 ASC

Reviewer 3: ```suggestion
SELECT
  "public"."rasa_sender"."channel" AS "channel",
  "public"."rasa_sender"."first_seen" AS "timestamp",
  count(distinct "public"."rasa_sender"."sender_key") AS "count"
FROM "public"."rasa_sender"
GROUP BY 1, 2
ORDER BY 1 ASC, 2 ASC
```

Reviewer 3: this attribute moved from the message to the sender table

Reviewer 3: and this is the number of "users" for this rather than sessions, sessions requires a join on the session table. what do we want to show, users per channel or sessions per channel @chandrikas ?

Reviewer 2: users per channel is fine - I think that is a more understandable and universal concept. Sessions is a little more technical

Author: 👍🏻

```

<figure align="center">
  <img
    alt="Number of sessions per channel visualised in Metabase."
    src={useBaseUrl("/img/analytics/graph-number-sessions-channel.png")}
    width="100%"
  />
  <figcaption>Number of sessions per channel visualised in Metabase.</figcaption>
</figure>

## Top N intents

> To improve your assistant, you could looking into the variety of intents your users express.

Reviewer: tiny typos 😄 
```suggestion
To improve your assistant, you could look into the variety of intents your users express.
```

GROUP BY 1, 2
ORDER BY 1 ASC, 2 ASC
```

<figure align="center">
  <img
    alt="Intent distribution over time visualised in Metabase."
    src={useBaseUrl("/img/analytics/graph-intent-distribution.png")}
    width="100%"
  />
  <figcaption>Intent distribution over time visualised in Metabase.</figcaption>
</figure>

## Escalation rate

Escalation rate or human hand-off rate is a measure of how many conversations does the assistant pass to a human agent.
This metric can help you gain a better understanding of what happens during a conversation.
Let's say you have an intent named `handoff_to_support`. You'll get the escalation rate over time
with this sample query:

```sql
WITH "sessions" AS (
    SELECT
        "public"."rasa_user_message"."session_id" AS "session_id",
        date_trunc('month', "public"."rasa_user_message"."timestamp") AS "timestamp",
        (
          CASE "public"."rasa_user_message"."intent"
            WHEN 'handoff_to_support'
            THEN 1 ELSE 0
          END
        ) AS "has_handoff_to_support"
    FROM "public"."rasa_user_message"
),
"sessions_with_handoff" AS (
    SELECT
      "session_id",
      "timestamp",
      SUM(

>         CASE

Reviewer 3: do we need the case here? isn't `has_handoff_to_support` set to `0` or `1`?

Reviewer 3: i think you did it differently for the abandonment one

Author: the thing is that for some sessions there was multiple "out_of_scope" intents. But yeah you're right I'll fix that
